### PR TITLE
Skip mame bios/devices

### PR DESCRIPTION
**Pull request by raelgc**
New attempt as now mame files were moved to external files.

### CODE REVIEW

**Comment by raelgc** [comment]
@joolswills New attempt after the my old PR #329, when mame stuff was hardcoded.

**Comment by joolswills** [comment]
I'm happy to merge this if others are.

**Comment by tomaz82** [comment]
I'm not, first off it has changes that is pure code style changes, to one that is not matching how ES style is.
It has missing newline at the end of both xml files, the fast manual search function that is used for the mamenames is not used for the asset, it uses the slower std::find. And that's just the ones I found after a quick 20 second look.

I would advice against a merge until all these issues are adressed.

**Comment by joolswills** [comment]
Sorry. I missed the styling issues - I agree with your comments @tomaz82

File: es-app/src/SystemData.cpp [hunk 1]
```diff
@@ -112,7 +112,8 @@ void SystemData::populateFolder(FileData* folder)
 				continue;
 
 			FileData* newGame = new FileData(GAME, filePath, mEnvData, this);
-			folder->addChild(newGame);
+			if(!newGame->isArcadeAsset())
+			  folder->addChild(newGame);
```
Comment (joolswills) [inline es-app/src/SystemData.cpp]: indent

File: es-core/src/MameNames.cpp [hunk 2]
```diff
@@ -36,27 +37,59 @@ MameNames* MameNames::getInstance()
 
 MameNames::MameNames()
 {
+	
+	// Read game names
 	std::string xmlpath = ResourceManager::getInstance()->getResourcePath(":/mamenames.xml");
-
 	if(!Utils::FileSystem::exists(xmlpath))
```
Comment (joolswills) [inline es-core/src/MameNames.cpp]: unrelated whitespace changes.

File: resources/mamedevices.xml [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+<device>a2cffa02</device>
+<device>m1comm</device>
+<device>ql_trump256</device>
+<device>list.txt</device>
+<device>dms3d2kp</device>
+<device>ti99_fdc</device>
+<device>dmv_k235</device>
+<device>mc1502_rom</device>
+<device>ql_pcmlqdi</device>
+<device>gfxultrp</device>
```
Comment (joolswills) [inline resources/mamedevices.xml]: no newlines at eof in xmls

File: es-core/src/MameNames.h [hunk 4]
```diff
@@ -9,10 +9,11 @@ class MameNames
...
 
-	static void       init       ();
-	static void       deinit     ();
-	static MameNames* getInstance();
-	std::string       getRealName(const std::string& _mameName);
+	static void              init       ();
+	static void              deinit     ();
+	static MameNames*        getInstance();
+	const bool               isAsset    (const std::string& _mameName);
+	std::string              getRealName(const std::string& _mameName);
```
Comment (joolswills) [inline es-core/src/MameNames.h]: no need for styling changes here. if you think the indent is wrong, it can be done as a separate later commit.

**Comment by raelgc** [comment]
Oh my, for some reason atom gone crazy with tabs (it should replace by 2 spaces). Let me fix all pointed issues.

**Comment by joolswills** [comment]
Using tabs is correct for the ES sources.

**Comment by raelgc** [comment]
While working in the fix, I can see the existing [mamenames.xml](https://raw.githubusercontent.com/RetroPie/EmulationStation/master/resources/mamenames.xml) has no newline at eof. Should I add to it too?

**Comment by raelgc** [comment]
@tomaz82 Why not, then, use the standard `std::binary_search` instead of a manual search?

**Comment by raelgc** [comment]
Ok, I've made all requested fixes using `gedit` as editor.

**Comment by tomaz82** [comment]
@raelgc because that "standard" std::binary_search still requires algorithm included ( which most likely will include even more headers ) which is pointless when we have a fast search algorithm there already.

**Comment by raelgc** [comment]
But we already have `algorithm` included in other files. Why should I write my own binary search when the standard c++ has a good one, in a library already included in the system?

**Comment by tomaz82** [comment]
I never said you should write your own, I said you should use the one VERY similar to the one we have in getRealName, it's your way of thinking about code that had ES take 50+ minutes to compile on the RPi3 before I optimized it down to less than 10 minutes.

**Comment by raelgc** [comment]
Interesting @tomaz82, I was not aware of the compilation time. Thanks for explaining. It's a really fare argument!

In defense of the `algorithm`, I just used because it was used in other files. But following your advice, I'll replace by a manual a binary search. For this, I'll need to sort the Mame assets vector (as I'm combining the `bios` xml and the `devices` xml). To avoid `std::sort`, should I merge them in just one file?

Just to share: for someone using Atom and the breaking styling: I gone to **Preferences** > **Editor** and disable `Atomic Soft Tabs` (while moving the cursor throw white-spaces it behaves like Tabs) and `Soft Tabs` (it tries to guess when use tabs vs spacing, but for me it was replacing the tabs for white-spaces).

**Comment by tomaz82** [comment]
@raelgc I understand, but what other files do, doesn't affect the compiletime of this file.
Looking at it more closely, I'd keep the 2 xml's entirely separated, even when stored as vectors, and make 2 functions ( isDevice and isBios ) instead of the isAsset,

**Comment by raelgc** [comment]
@tomaz82 Sure, I'll work on this.

**Comment by raelgc** [comment]
@pjft Do you mind to test and give any feedback? 

Expected result is to **not** see mame devices in the gamelist, by example, `neogeo.zip` in mame/neogeo/arcade gamelist.

**Comment by pjft** [comment]
Hi. I don't have much time this week, apologies. I'll give it a go when I have the chance, but I suspect the main feedback here/green light would have to come from @tomaz82 as he's the one who had specific concerns about the code and libraries being used. I see you're still including <algorithm>, which I believe was one of the concerns @tomaz82 raised, unless I missed something else here.

**Comment by tomaz82** [comment]
I'm still waiting for the requested changes

**Comment by raelgc** [comment]
@tomaz82 Just removed `algorithm` library. I've separated the 2 lists like you requested in a previous commit.

**Comment by pjft** [comment]
So, I had a few moments today and compiled this and tested.

Still, the BIOS files still show up on my arcade set. Am I meant to enable one option? At least neogeo.zip and pgm.zip show up.

**Comment by raelgc** [comment]
@pjft Any chance we have different setups? I tested here with `neogeo.zip` in both `neogeo` and `arcade` folders and it was not displayed, while with the `master` branch it was displayed.

**Comment by pjft** [comment]
Maybe. The only thing is that mine are in my gamelists - is that a thing for you?

**Comment by raelgc** [comment]
@pjft Nice catch! Indeed, not in my gamelist! Maybe I should implement the check in more places...

**Comment by pjft** [comment]
Well, try adding it to your gamelist and see if still shows :) Maybe that's it, maybe not :)

**Comment by raelgc** [comment]
Yes, it's it (just checked). Thanks for that!

I'll check where the gamelist is parsed and run the same skip logic.

**Comment by pjft** [comment]
Well, see if it could be applied in a single place then - I'm not especially fond of the idea of duplicating logic unless really necessary. 

Thanks.

**Comment by raelgc** [comment]
@pjft I just pushed a change to prevent arcade assets get read from gamelist.xml. When you have some free time, please, can you test again?

**Comment by raelgc** [comment]
@jrassa Just made the requested changes.

**Comment by pjft** [comment]
Hi!

I'm so sorry for the delay, and also for the update. For some reason, they are still showing.

I'm running the binary from commit 

baa9cff697919f85571da6c22f3d0fd07c62f059

from 2 days ago :/

The resources from the gamelists are still showing somehow - at least neogeo.zip and pgm.zip .

**Comment by raelgc** [comment]
@pjft Sounds like I'm still missing some other entry points. 

Any chance your ES is configured to something different than the default? I mean, a filter applied from our tests on the improved Kids mode, something like that?

**Comment by pjft** [comment]
I don't expect that to be the case, but I'll look into it in the coming days and get back to you.

**Comment by raelgc** [comment]
@pjft Just to share: I tested both cases (mame assets outside the gamelist), and later manually added assets in the `gamelist.xml`. Both with a fresh retropie install (and they worked) in 2 different machines. 

Just wondering which is the difference between our setups. For this reason I'm mentioning the filters. But of course, can be other features.

File: es-app/src/Gamelist.cpp [hunk 5]
```diff
@@ -48,7 +48,12 @@ FileData* findOrCreateFile(SystemData* system, const std::string& path, FileType
 			}
 
 			FileData* file = new FileData(type, path, system->getSystemEnvData(), system);
-			treeNode->addChild(file);
+
+			// skipping arcade assets from gamelist
+			if(!treeNode->isArcadeAsset())
```
Comment (pjft) [inline es-app/src/Gamelist.cpp]: I can confirm that in my scenario, it does not go through here.

File: es-app/src/SystemData.cpp [hunk 6]
```diff
@@ -112,8 +112,13 @@ void SystemData::populateFolder(FileData* folder)
 				continue;
 
 			FileData* newGame = new FileData(GAME, filePath, mEnvData, this);
-			folder->addChild(newGame);
-			isGame = true;
+
+			// preventing new arcade assets to be added
+			if(!newGame->isArcadeAsset())
```
Comment (pjft) [inline es-app/src/SystemData.cpp]: I can confirm it goes through here, but doesn't get added here.

File: es-app/src/Gamelist.cpp (hunk 5 above)
Comment (pjft) [inline es-app/src/Gamelist.cpp]: Here's your error. treeNode will be the parent folder in this case. You probably want to run this check against "file", which is the newly-created file.

That being said, I'm not fully comfortable with the idea of even creating a FileData object if we're going to discard it - I'd rather only create it if we're going to use it, otherwise make sure to return NULL - and see if we don't crash.

Comment (raelgc) [reply es-app/src/Gamelist.cpp]: @pjft First: thank you for your time and patience.

I just pushed the requested changes but not yet the `NULL` return (before double check with you).

If I return NULL, it'll not crash, but it'll produce this output:

> lvl0: 	Error finding/creating FileData for "/home/rael/RetroPie/roms/neogeo/neogeo.zip", skipping.`

Returning the `FileData`, no error output.

How should I proceed?

Comment (pjft) [reply es-app/src/Gamelist.cpp]: I'd leave a definitive answer for others to chime in on, but in my perspective I'm not wholly bothered by that warning. The alternative is to create FileData objects that are dangling without a parent, take up memory, and never get explicitly removed. I'd imagine there'd be some edge case where that'd come and bite us back.

But if anyone else has any recommendation or preference, I'd happily take that.

File: es-app/src/SystemData.cpp (hunk 6 above)
Comment (raelgc) [reply es-app/src/SystemData.cpp]: Just to answer your comment: here is the place where files not in `gamelist.xml` are added.

**Comment by raelgc** [comment]
@pjft Sorry for the long time away from this. But how about now? I'm avoiding the error message and avoiding create any additional metadata with the `FileData` when it's an arcade asset.

**Comment by pjft** [comment]
Seems to work well, from my brief testing.

**Comment by raelgc** [comment]
@joolswills: in the last weeks, I just got playing with settings and thinking about you've said about performance and focus on Raspberry devices. And probably today the only "annoying" BIOS we have displayed in the gamelist is usually `neogeo.zip`, which can be moved to the `~/Retropie/BIOS` folder (at least for `lr-fbalpha`).

So, in a second moment, I'm no more inclined to include all this overhead that my PR will produce just to handle few files we can easily get hidden.

What do you think?

**Comment by dankcushions** [comment]
i still think the PR is a good idea. all other ways of hiding BIOS files need manual intervention from the user.

neogeo.zip is needed in the roms folder for MAME cores, and it's not the only BIOS file needed in MAME - 2003 also needs skns, pgm, and taitofx1. later versions of MAME will have need even more BIOS files.

fbalpha has neogeo.zip (which can be hidden via the BIOS folder, but that requires manual intervention by the user, and knowledge of that niche behaviour), but it also requires isgsm, nmk004, pgm, skns, and ym2608.

**Comment by raelgc** [comment]
Well, in any case, I just squashed the commits.

**Comment by joolswills** [comment]
I'm going to merge this unless anyone has an issue

**Comment by joolswills** [comment]
Thanks for your work on this.
